File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    void solve(int idx, vector<int>& candidates, vector<int>& picks,vector<vector<int>>& ans,int target){
        int n = candidates.size();
        if(target==0) {
            ans.push_back(picks);
            return;
        }
        if(idx==n) return;
        
        for(int i=idx;i<n;i++){
            if(candidates[i]>target) break;
            if(i>idx&&candidates[i]==candidates[i-1]) continue;
            picks.push_back(candidates[i]);
            solve(i+1,candidates,picks,ans,target-candidates[i]);
            picks.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(candidates.begin(),candidates.end());
        solve(0,candidates,temp,ans,target);
        return ans;
    }
};

File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        ctr = Counter(candidates)
        uniques = list(ctr.keys())
        N,ans = len(uniques),[]
        def solve(idx,target,path=[]):
            # if idx==N:
            if target==0:
                ans.append(list(path))
            else:
                if idx>=N:return
                cur = uniques[idx]
                for pickCount in range(0,ctr[cur]+1):
                    if pickCount*cur<=target:
                        solve(idx+1,target-pickCount*cur,path+[cur]*pickCount)
        solve(0,target,[])
        return ans
    
#         candidates.sort()
#         N = len(candidates)
#         ans = []
#         def solve(idx,target,path):
#             # if idx==N:
#             if target==0:
#                 ans.append(list(path))
#             else:
#                 for i in range(idx,N):
#                     if i>idx and candidates[i]==candidates[i-1]:continue
#                     if candidates[i]>target:break
#                     path.append(candidates[i])
#                     solve(i+1,target-candidates[i],path)
#                     path.pop()
                    
                    
#         solve(0,target,[])
#         return ans